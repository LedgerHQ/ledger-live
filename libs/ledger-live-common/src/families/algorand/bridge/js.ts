import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import type { AlgorandTransaction } from "../types";
import { sync, scanAccounts } from "../js-synchronization";
import {
  createTransaction,
  prepareTransaction,
} from "../js-prepareTransaction";
import { estimateMaxSpendable } from "../js-estimateMaxSpendable";
import { getTransactionStatus } from "../js-getTransactionStatus";
import { signOperation } from "../js-signOperation";
import { broadcast } from "../js-broadcast";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const updateTransaction = (t, patch) => {
  return { ...t, ...patch };
};

const preload = async () => Promise.resolve({});

const hydrate = () => {};

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<AlgorandTransaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast,
  estimateMaxSpendable,
};

export default {
  currencyBridge,
  accountBridge,
};
