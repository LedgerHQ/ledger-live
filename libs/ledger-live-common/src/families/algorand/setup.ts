// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
