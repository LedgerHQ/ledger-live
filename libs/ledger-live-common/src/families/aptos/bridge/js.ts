import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import { getSerializedAddressParameters } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../synchronisation";
import getTransactionStatus from "../getTransactionStatus";
import prepareTransaction from "../prepareTransaction";
import createTransaction from "../createTransaction";
import estimateMaxSpendable from "../estimateMaxSpendable";
import signOperation from "../signOperation";
import broadcast from "../broadcast";

const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};

const updateTransaction = (t: Transaction, patch: Partial<Transaction>): Transaction => ({
  ...t,
  ...patch,
});

const receive = makeAccountBridgeReceive();

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  getSerializedAddressParameters,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
