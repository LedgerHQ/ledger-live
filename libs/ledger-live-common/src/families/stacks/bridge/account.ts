import { Account, AccountBridge } from "@ledgerhq/types-live";
import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { getTransactionStatus } from "../getTransactionStatus";
import { estimateMaxSpendable } from "../estimateMaxSpendable";
import { prepareTransaction } from "../prepareTransaction";
import { Transaction, TransactionStatus } from "../types";
import { createTransaction } from "../createTransaction";
import { signOperation } from "../signOperation";
import { sync } from "../synchronization";
import { broadcast } from "../broadcast";

const receive = makeAccountBridgeReceive();

export const accountBridge: AccountBridge<Transaction, Account, TransactionStatus> = {
  createTransaction,
  updateTransaction: defaultUpdateTransaction,
  getTransactionStatus,
  prepareTransaction,
  estimateMaxSpendable,
  signOperation,
  sync,
  receive,
  broadcast,
};
