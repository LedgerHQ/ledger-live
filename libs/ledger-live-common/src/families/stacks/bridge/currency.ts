import { makeScanAccounts } from "../../../bridge/jsHelpers";
import { getAccountShape } from "./utils/utils";
import { CurrencyBridge } from "@ledgerhq/types-live";

const scanAccounts = makeScanAccounts({ getAccountShape });

export const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};
