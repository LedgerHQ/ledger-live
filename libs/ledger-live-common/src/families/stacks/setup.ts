// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-stacks";
import stacksResolver, { signMessage } from "@ledgerhq/coin-stacks/signer/index";
import makeCliTools from "@ledgerhq/coin-stacks/test/cli";
import Transport from "@ledgerhq/hw-transport";
import { Bridge } from "@ledgerhq/types-live";
import BlockstackApp from "@zondax/ledger-stacks";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";
import { StacksSigner, Transaction } from "./types";

const createSigner: CreateSigner<StacksSigner> = (transport: Transport) => {
  return new BlockstackApp(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner));

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, stacksResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
