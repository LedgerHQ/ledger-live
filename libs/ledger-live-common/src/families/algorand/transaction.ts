import { BigNumber } from "bignumber.js";
import type {
  AlgorandAccount,
  AlgorandTransaction,
  AlgorandTransactionRaw,
} from "./types";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon as fromTransactionStatusRaw,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon as toTransactionStatusRaw,
} from "../../transaction/common";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
export const formatTransaction = (
  {
    mode,
    subAccountId,
    amount,
    recipient,
    fees,
    useAllAmount,
  }: AlgorandTransaction,
  mainAccount: AlgorandAccount
): string => {
  const account =
    (subAccountId &&
      (mainAccount.subAccounts || []).find((a) => a.id === subAccountId)) ||
    mainAccount;
  return `
    ${
      mode === "claimReward"
        ? "CLAIM REWARD"
        : mode === "optIn"
        ? "OPT_IN"
        : "SEND"
    } ${
    useAllAmount
      ? "MAX"
      : formatCurrencyUnit(getAccountUnit(account), amount, {
          showCode: true,
          disableRounding: false,
        })
  }
    TO ${recipient}
    with fees=${
      !fees
        ? "?"
        : formatCurrencyUnit(getAccountUnit(mainAccount), fees, {
            showCode: true,
            disableRounding: false,
          })
    }`;
};

const fromTransactionRaw = (
  tr: AlgorandTransactionRaw
): AlgorandTransaction => {
  const common = fromTransactionCommonRaw(tr);
  return {
    ...common,
    family: tr.family,
    fees: tr.fees ? new BigNumber(tr.fees) : null,
    memo: tr.memo,
    mode: tr.mode,
    assetId: tr.assetId,
  };
};

const toTransactionRaw = (t: AlgorandTransaction): AlgorandTransactionRaw => {
  const common = toTransactionCommonRaw(t);
  return {
    ...common,
    family: t.family,
    fees: t.fees ? t.fees.toString() : null,
    memo: t.memo,
    mode: t.mode,
    assetId: t.assetId,
  };
};

export default {
  formatTransaction,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
  formatTransactionStatus,
};
