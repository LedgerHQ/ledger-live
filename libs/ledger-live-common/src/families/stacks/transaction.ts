import BigNumber from "bignumber.js";
import { StacksNetwork } from "./bridge/utils/api.types";
import type { Account } from "@ledgerhq/types-live";

import type { Transaction, TransactionRaw } from "./types";
import { formatTransactionStatus } from "@ledgerhq/coin-framework/formatters";
import {
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon as fromTransactionStatusRaw,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon as toTransactionStatusRaw,
} from "@ledgerhq/coin-framework/serialization";
import { getAccountCurrency } from "../../account";
import { formatCurrencyUnit } from "../../currencies";

export const formatTransaction = (
  { recipient, useAllAmount, amount }: Transaction,
  account: Account,
): string => `
SEND ${
  useAllAmount
    ? "MAX"
    : amount.isZero()
      ? ""
      : " " +
        formatCurrencyUnit(getAccountCurrency(account).units[0], amount, {
          showCode: true,
          disableRounding: true,
        })
}
TO ${recipient}`;

export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);

  // validate if network is valid
  if (!StacksNetwork[tr.network]) {
    throw new Error(`network ${tr.network} not valid`);
  }

  return {
    ...common,
    family: tr.family,
    nonce: tr.nonce !== undefined ? new BigNumber(tr.nonce) : undefined,
    fee: tr.fee !== undefined ? new BigNumber(tr.fee) : undefined,
    amount: new BigNumber(tr.amount),
    network: tr.network as keyof typeof StacksNetwork,
    anchorMode: tr.anchorMode,
    memo: tr.memo,
  };
};

const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);

  return {
    ...common,
    family: t.family,
    fee: t.fee !== undefined ? t.fee.toFixed() : undefined,
    nonce: t.nonce !== undefined ? t.nonce.toFixed() : undefined,
    amount: t.amount.toFixed(),
    network: t.network,
    anchorMode: t.anchorMode,
    memo: t.memo,
  };
};

export default {
  formatTransaction,
  fromTransactionRaw,
  toTransactionRaw,
  formatTransactionStatus,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
};
