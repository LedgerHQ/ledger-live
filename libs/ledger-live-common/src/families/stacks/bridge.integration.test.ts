import { CurrenciesData } from "@ledgerhq/types-live";
import { AnchorMode } from "@stacks/transactions";
import type { DatasetTest } from "@ledgerhq/types-live";
import { BigNumber } from "bignumber.js";
import {
  AmountRequired,
  InvalidAddressBecauseDestinationIsAlsoSource,
  NotEnoughBalance,
} from "@ledgerhq/errors";

import "../../__tests__/test-helpers/setup";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "./transaction";
import { testBridge } from "../../__tests__/test-helpers/bridge";

const SEED_IDENTIFIER = "SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9";
const SEED_IDENTIFIER_PUBKEY = "022a460decc9dba8c452927fecb33d7ae25a8d79dc5442b84feaf8f3aa0e2b575d";
const ACCOUNT_1 = "SP2DV2RVZP1A69Q6VAG5PHEQ6ZHQHZPCV84TMYNGN";

const stacks: CurrenciesData<Transaction> = {
  FIXME_ignoreAccountFields: [],
  IgnorePrepareTransactionFields: ["fee"],
  scanAccounts: [
    {
      name: "stacks seed 1",
      apdus: `
      => 09010016142c0000807d160080000000800000000000000000
      <= 022a460decc9dba8c452927fecb33d7ae25a8d79dc5442b84feaf8f3aa0e2b575d5350334b5337564d59325a4e45365342383850485234534b524b324545504853384e384d43434252399000
      => 09010016142c0000807d160080000000800000000001000000
      <= 03a2d4bbb54c2f44c3a205b15ca7d2049a2bb66e50e100304f932b3b99ed3a99cb5350324136463652374632395a3338533259434a4d37484a44564532345332564757543542545a50419000
      => 09010016142c0000807d160080010000800000000000000000
      <= 034bbb21911e9d4502ac170162c5c05cbba65cbb79e5b46adf95152af2d0e78a4553503244563252565a503141363951365641473550484551365a4851485a5043563834544d594e474e9000
      => 09010016142c0000807d160080020000800000000000000000
      <= 020637c4d824458f6a11551f51df9e5be9e96c2e3a2e5bdb25d2916cc1dd5ba2d753504e5a4e434333433244525450574a4a384e534d484d575744304643415459413843565a3834459000
      => 09010016142c0000807d160080030000800000000000000000
      <= 02e57bb8c28c4c768cf65f8e28935482e40afbb33168d26810199f39a719cc05f0535031564d30584546574251345a543838384654333447333047474635324459395951575a434151309000
      => 09010016142c0000807d160080040000800000000000000000
      <= 0264b6d44f720fa50055dc294276bf15ce22e4f3dec48510b317ab8c07419c1b7753503248574a58594e4747594354535952545036323636424a483858565a4759504d455248514538529000
      `,
    },
  ],
  accounts: [
    {
      raw: {
        id: `js:2:stacks:${SEED_IDENTIFIER_PUBKEY}:`,
        seedIdentifier: SEED_IDENTIFIER,
        name: "Stacks 1",
        derivationMode: "",
        index: 0,
        freshAddress: SEED_IDENTIFIER,
        freshAddressPath: "",
        blockHeight: 0,
        operations: [],
        xpub: SEED_IDENTIFIER_PUBKEY,
        pendingOperations: [],
        currencyId: "stacks",
        lastSyncDate: "",
        balance: "1000",
      },
      transactions: [
        {
          name: "Source and destination are the equal",
          transaction: fromTransactionRaw({
            family: "stacks",
            nonce: "1",
            network: "mainnet",
            anchorMode: AnchorMode.Any,
            recipient: SEED_IDENTIFIER,
            amount: "1",
          }),
          expectedStatus: {
            errors: {
              recipient: new InvalidAddressBecauseDestinationIsAlsoSource(),
            },
            warnings: {},
          },
        },
        {
          name: "Not enough balance",
          transaction: fromTransactionRaw({
            family: "stacks",
            nonce: "1",
            network: "mainnet",
            anchorMode: AnchorMode.Any,
            recipient: ACCOUNT_1,
            amount: "10000000000",
          }),
          expectedStatus: {
            errors: {
              amount: new NotEnoughBalance(),
            },
            warnings: {},
          },
        },
        {
          name: "Amount Required",
          transaction: fromTransactionRaw({
            family: "stacks",
            nonce: "1",
            network: "mainnet",
            anchorMode: AnchorMode.Any,
            recipient: ACCOUNT_1,
            amount: "0",
          }),
          expectedStatus: {
            errors: {
              amount: new AmountRequired(),
            },
            warnings: {},
          },
        },
        {
          name: "New account and sufficient amount",
          transaction: fromTransactionRaw({
            family: "stacks",
            nonce: "1",
            network: "mainnet",
            anchorMode: AnchorMode.Any,
            recipient: ACCOUNT_1,
            amount: "1",
          }),
          expectedStatus: {
            amount: new BigNumber("1"),
            errors: {},
            warnings: {},
          },
        },
      ],
    },
  ],
};

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    stacks,
  },
};

testBridge(dataset);
