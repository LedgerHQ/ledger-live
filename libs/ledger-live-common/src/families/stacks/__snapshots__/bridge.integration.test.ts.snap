// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stacks currency bridge scanAccounts stacks seed 1 1`] = `
Array [
  Object {
    "balance": "100000",
    "currencyId": "stacks",
    "derivationMode": "",
    "freshAddress": "SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9",
    "freshAddressPath": "44'/5757'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9",
        "derivationPath": "44'/5757'/0'/0/0",
      },
    ],
    "id": "js:2:stacks:SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9:",
    "index": 0,
    "name": "Stacks 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "022a460decc9dba8c452927fecb33d7ae25a8d79dc5442b84feaf8f3aa0e2b575d",
    "spendableBalance": "100000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 6,
    "used": true,
    "xpub": "022a460decc9dba8c452927fecb33d7ae25a8d79dc5442b84feaf8f3aa0e2b575d",
  },
  Object {
    "balance": "0",
    "currencyId": "stacks",
    "derivationMode": "",
    "freshAddress": "SP2DV2RVZP1A69Q6VAG5PHEQ6ZHQHZPCV84TMYNGN",
    "freshAddressPath": "44'/5757'/1'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "SP2DV2RVZP1A69Q6VAG5PHEQ6ZHQHZPCV84TMYNGN",
        "derivationPath": "44'/5757'/1'/0/0",
      },
    ],
    "id": "js:2:stacks:SP2DV2RVZP1A69Q6VAG5PHEQ6ZHQHZPCV84TMYNGN:",
    "index": 1,
    "name": "Stacks 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "022a460decc9dba8c452927fecb33d7ae25a8d79dc5442b84feaf8f3aa0e2b575d",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 6,
    "used": false,
    "xpub": "034bbb21911e9d4502ac170162c5c05cbba65cbb79e5b46adf95152af2d0e78a45",
  },
]
`;

exports[`stacks currency bridge scanAccounts stacks seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:stacks:SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9:",
      "blockHash": null,
      "blockHeight": 76373,
      "contract": undefined,
      "extra": Object {},
      "fee": "180",
      "hash": "0x6f1075f275482ca923b2fe9f8386c50fde0d3bdf5fee880459c6009d19777ffb",
      "id": "js:2:stacks:SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9:-0x6f1075f275482ca923b2fe9f8386c50fde0d3bdf5fee880459c6009d19777ffb-IN",
      "operator": undefined,
      "recipients": Array [
        "SP3KS7VMY2ZNE6SB88PHR4SKRK2EEPHS8N8MCCBR9",
      ],
      "senders": Array [
        "SP1QM16S93TEMM1Q9QHB8WV1EEEC3Z2Y3P0AE287S",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "100000",
    },
  ],
  Array [],
]
`;
